source/MaterialXGenShader/ShaderGraph.cpp

```diff
@@ -757,7 +766,78 @@ ShaderNode* ShaderGraph::createNode(ConstNodePtr node, GenContext& context)
     // Apply color and unit transforms to each input.
     applyInputTransforms(node, newNode, context);
 
-    return newNode.get();
+    return newNode;
+}
+
+// Insert a new node between the output of the graph and its upstream connection, reconnecting the upstream to the specified input on
+// the new node, if present.
+ShaderNode* ShaderGraph::inlineNodeBeforeOutput(ShaderGraphOutputSocket* output, const std::string& newNodeName, const std::string& nodeDefName, const std::string& inputName, const std::string& outputName, GenContext& context)
+{
+    auto nodeDef = _document->getNodeDef(nodeDefName);
+    if (!nodeDef)
+    {
+        throw ExceptionShaderGenError("Cannot find NodeDef '"+nodeDefName+"' when inserting node '"+newNodeName+"'");
+    }
+
+    // Check to see if the nodedef has the specified input/output ports
+    OutputPtr nodeDefOutput = nodeDef->getOutput(outputName);
+    if (!nodeDefOutput)
+    {
+        throw ExceptionShaderGenError("Output '"+outputName+"' not found on NodeDef '"+nodeDefName+"'");
+    }
+
+    InputPtr nodeDefInput = nullptr;
+    if (!inputName.empty())
+    {
+        // only look for the input if we are given an inputName - its valid to insert the node
```

Comment:
Minor note: These new comments should begin with capitalized words for consistency.